torchmultimodal/models/gpt.py

```diff
@@ -28,12 +28,221 @@ class TransformerLayerOutput(NamedTuple):
+            embeddings for the input modality. Defaults to ``None``.
+        out_modality (Tensor, optional): Tensor of dimension ``(b, out_seq_len, c')`` containing tokenized
+            embeddings for the output modality. Defaults to ``None``.
+        in_pos_ids (Tensor, optional): Tensor of dimension ``(b, in_seq_len)`` containing indices for the
+            input modality position embeddings. Defaults to ``None``.
+        out_pos_ids (Tensor, optional): Tensor of dimension ``(b, out_seq_len)`` containing indices for the
+            output modality position embeddings. Defaults to ``None``.
+        attn_mask (Tensor, optional): Tensor of dimension ``(b, total_seq_len, total_seq_len)``. Contains 1s for
+            positions to attend to and 0s for masked positions. Defaults to ``None``.
+        head_mask (Tensor, optional): Tensor of dimension ``(b, h, total_seq_len, total_seq_len)``. Contains 1s
+            for attention heads to use and 0s for masked heads. Defaults to ``None``.
+        use_cache (bool, optional): If ``True``, caches past key/value tensors for faster decoding. If ``False``,
+            recomputes key and value for each decoding step. Defaults to ``False``.
+        causal (bool. optional): If ``True``, use causal attention. Defaults to ``False``.
+        right_shift (bool): If ``True``, shifts the embedding vectors to the right and prepends it with start of
+            sentence token. Defaults to ``False``. This option is disregarded during training mode
+        return_attn_weights (bool, optional): If ``True``, returns attention probabilities of each transformer
+            layer. Defaults to ``False``.
+        return_hidden_states (bool): If ``True``, returns the embeddings of each transformer layer. Defaults to
+            ``False``.
+
+    Raises:
+        AttributeError: If input tokenizer does not implement methods ``encode`` and ``decode`` or output
+            tokenizer does not implement methods ``encode``, ``decode`` and ``lookup``.
+    """
+
+    def __init__(
+        self,
+        d_model: int,
+        num_tokens: int,
+        in_tokenizer: nn.Module,
+        out_tokenizer: nn.Module,
+        mm_decoder: nn.Module,
+    ) -> None:
+        super().__init__()
+        if not all(
+            [hasattr(in_tokenizer, attr_name) for attr_name in ["encode", "decode"]]
+        ):
+            raise AttributeError(
+                "Input modality tokenizer must have methods 'encode', 'decode' and 'lookup'."
+            )
+
+        if not all(
+            [
+                hasattr(out_tokenizer, attr_name)
+                for attr_name in ["encode", "decode", "lookup"]
+            ]
+        ):
+            raise AttributeError(
+                "Output modality tokenizer must have methods 'encode', 'decode' and 'lookup'."
+            )
+
+        self.d_model = d_model
+        self.num_tokens = num_tokens
```

Comment:
I don't see `num_tokens` and `d_model` used elsewhere except for initializing the head, do we need to keep them as attributes?

Reply:
They are used by `LayerNorm` and `to_logit`: 

```
self.norm = nn.LayerNorm(d_model)
self.to_logit = nn.Linear(d_model, num_tokens, bias=False)
self.to_logit.weight.data.copy_(torch.zeros(self.num_tokens, self.d_model))
```

Reply:
right, I was just thinking you could use `d_model` and `num_tokens` directly and retire `self.d_model` and `self.num_tokens`